blockprod/src/detail/mod.rs

```diff
@@ -443,11 +346,129 @@ impl BlockProduction {
                     };
 
                     let block = Block::new_from_header(signed_block_header, block_body.clone())?;
-                    return Ok((block, end_confirm_receiver));
+                    return Ok((block, job_finished_receiver));
                 }
             }
         }
     }
+
+    #[allow(clippy::too_many_arguments)]
+    fn spawn_block_solver(
+        &self,
+        current_tip_index: &GenBlockIndex,
+        stop_flag: Arc<AtomicBool>,
+        block_body: &BlockBody,
+        timestamp: BlockTimestamp,
+        finalize_block_data: Option<FinalizeBlockInputData>,
+        consensus_data: ConsensusData,
+        ended_sender: mpsc::Sender<()>,
+        result_sender: oneshot::Sender<Result<SignedBlockHeader, BlockProductionError>>,
+    ) -> Result<(), BlockProductionError> {
+        self.mining_thread_pool.spawn({
+            let chain_config = Arc::clone(&self.chain_config);
+            let current_tip_height = current_tip_index.block_height();
+            let stop_flag = Arc::clone(&stop_flag);
+
+            let merkle_proxy =
+                block_body.merkle_tree_proxy().map_err(BlockCreationError::MerkleTreeError)?;
+
+            let mut block_header = BlockHeader::new(
+                current_tip_index.block_id(),
+                merkle_proxy.merkle_tree().root(),
+                merkle_proxy.witness_merkle_tree().root(),
+                timestamp,
+                consensus_data,
+            );
+
+            move || {
+                let signed_block_header = consensus::finalize_consensus_data(
+                    &chain_config,
+                    &mut block_header,
+                    current_tip_height,
+                    stop_flag,
+                    finalize_block_data,
+                )
+                .map_err(BlockProductionError::FailedConsensusInitialization);
+
+                let _ended_sender = OnceDestructor::new(move || {
+                    // This can fail if the function exited before the mining thread finished
+                    let _send_whether_ended = ended_sender.send(());
+                });
+
+                result_sender
+                    .send(signed_block_header)
+                    .expect("Failed to send block header back to main thread");
+            }
+        });
+
+        Ok(())
+    }
+}
+
+fn generate_finalize_block_data(
+    chain_config: &ChainConfig,
+    chainstate_handle: &(dyn ChainstateInterface),
+    best_block_index: &GenBlockIndex,
+    block_height: BlockHeight,
+    sealed_epoch_index: Option<EpochIndex>,
+    sealed_epoch_randomness: Option<PoSRandomness>,
+    input_data: GenerateBlockInputData,
+) -> Result<Option<FinalizeBlockInputData>, ConsensusPoSError> {
+    match input_data {
+        GenerateBlockInputData::PoS(pos_input_data) => {
+            let sealed_epoch_index = sealed_epoch_index.ok_or(ConsensusPoSError::NoEpochData)?;
```

Comment:
Btw, this is wrong. It just caused an error for Pavel while trying to stake. Starting from nothing, this won't exist. So it should be tolerated.

Reply:
For Pavel I just recommended using `unwrap_or(0)`. See if you can fix this in a way where this can be just skipped. I don't like the idea that randomness is separate from this. They both should either exist together or not, where genesis uses initial randomness.

Also another thing, there's no reason for randomness to be an Option. It should always exist. I see you have a TODO there though. But for this function, there should always be a source of randomness, whether it's from genesis or otherwise. There's no situation where randomness won't exist.

Reply:
we could add initial epoch data in `process_genesis`

Reply:
That's a good idea, actually.